services/121-service/src/payments/fsp-integration/commercial-bank-ethiopia/commercial-bank-ethiopia.service.ts

```diff
@@ -0,0 +1,281 @@
+  implements FinancialServiceProviderIntegrationInterface
+{
+  @InjectRepository(RegistrationEntity)
+  private readonly registrationRepository: Repository<RegistrationEntity>;
+  @InjectRepository(TransactionEntity)
+  public transactionRepository: Repository<TransactionEntity>;
+  @InjectRepository(ProgramEntity)
+  public programRepository: Repository<ProgramEntity>;
+  @InjectRepository(ProgramFspConfigurationEntity)
+  public programFspConfigurationRepository: Repository<ProgramFspConfigurationEntity>;
+
+  public constructor(
+    private readonly commercialBankEthiopiaApiService: CommercialBankEthiopiaApiService,
+    private readonly transactionsService: TransactionsService,
+  ) {}
+
+  public async sendPayment(
+    paymentList: PaPaymentDataDto[],
+    programId: number,
+    paymentNr: number,
+  ): Promise<FspTransactionResultDto> {
+    const config = await this.programFspConfigurationRepository
+      .createQueryBuilder('fspConfig')
+      .select('name')
+      .addSelect('value')
+      .where('fspConfig.programId = :programId', { programId })
+      .andWhere('fsp.fsp = :fspName', { fspName: paymentList[0].fspName })
+      .leftJoin('fspConfig.fsp', 'fsp')
+      .getRawMany();
+
+    const credentials: { username: string; password: string } = {
+      username: config.find((c) => c.name === 'username')?.value,
+      password: config.find((c) => c.name === 'password')?.value,
+    };
+
+    const program = await this.programRepository.findOneBy({
+      id: programId,
+    });
+
+    const fspTransactionResult = new FspTransactionResultDto();
+    fspTransactionResult.paList = [];
+    fspTransactionResult.fspName = FspName.commercialBankEthiopia;
+
+    const referenceIds = paymentList.map((payment) => payment.referenceId);
+    const userInfo = await this.getUserInfo(referenceIds);
+
+    for (const payment of paymentList) {
+      const resultUser = await this.getObjectByReferenceId(payment, userInfo);
+
+      const payload = this.createPayloadPerPa(payment, resultUser, program);
+
+      const paymentRequestResultPerPa = await this.sendPaymentPerPa(
+        payload,
+        payment.referenceId,
+        credentials,
+      );
+      fspTransactionResult.paList.push(paymentRequestResultPerPa);
+      // Storing the per payment so you can continiously seed updates of transactions in HO-Portal
+      await this.transactionsService.storeTransactionUpdateStatus(
+        paymentRequestResultPerPa,
+        programId,
+        paymentNr,
+      );
+    }
+    return fspTransactionResult;
+  }
+
+  public async getObjectByReferenceId(
+    payment: any,
+    data: any,
+  ): Promise<
+    [
+      {
+        referenceId: string;
+        programname?: string;
+        fspname?: string;
+        value: string;
+        debitTheIrRef?: string;
+      },
+    ]
+  > {
+    const results = data.filter(
+      (item) => item.referenceId === payment.referenceId,
+    );
+
+    if (payment.transactionId) {
```

Comment:
@Vladimir-Petkov I'm not understanding where "transactionId" can end up to be an attribute of "payment", which is btw of the PaPaymentDataDto class, which doesn't have an attribute "transactionId".

Reply:
@Vladimir-Petkov in general I see a lot of 'any' or undefined classes for function input parameters. When I start to add some of those, the code starts giving errors (like for example above 'transactionId' not being part of 'PaPaymentDataDto', but there's more. Could you make sure to add proper dto classes for all input variables, and fix the resulting issues.

Reply:
@jannisvisser When you send retry payment, you have transactionId and I am using this to get debitTheIrRef, because we need that to check if transaction is duplicated or not.

Reply:
@jannisvisser Let me check the different classes and try to add dto classes.